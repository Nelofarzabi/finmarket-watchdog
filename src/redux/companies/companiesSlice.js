import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

const initialState = {
  companies: [],
  filteredCompanies: [],
  companyDetail: {},
  statement: [],
  loading: false,
  error: '',
  typing: false,
};
// const API_KEY = 'c0405893fd96337e82d58e3ee28f1c63';
const API_KEY = 'd0627ca47e8ec703f960f6fa495a3c45';

export const getActiveCompanies = createAsyncThunk('companies/getCompanies', async () => {
  try {
    const response = await fetch(`https://financialmodelingprep.com/api/v3/stock_market/actives?apikey=${API_KEY}`);
    const data = await response.json();
    return data;
  } catch (error) {
    return error;
  }
});
export const getCompanyDetail = createAsyncThunk('companies/getCompanyDetail', async (symbol) => {
  try {
    const response = await fetch(`https://financialmodelingprep.com/api/v3/profile/${symbol}?apikey=${API_KEY}`);
    const data = await response.json();
    return data;
  } catch (error) {
    return error;
  }
});
export const getCompanyStatement = createAsyncThunk('companies/getCompanyStatement', async (symbol) => {
  try {
    const response = await fetch(`https://financialmodelingprep.com/api/v3/income-statement/${symbol}?limit=120&apikey=${API_KEY}`);
    const data = await response.json();
    return data;
  } catch (error) {
    return error;
  }
});

const companiesSlice = createSlice({
  name: 'companies',
  initialState,
  reducers: {
    filterCompanies: (state, action) => {
      state.filteredCompanies = state.companies
        .filter((company) => company.name.toLowerCase()
          .startsWith(action.payload.toLowerCase()));
    },
    updateTyping: (state, action) => {
      state.typing = !!action.payload;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(getActiveCompanies.pending, (state) => {
        state.loading = true;
      })
      .addCase(getActiveCompanies.fulfilled, (state, action) => {
        state.loading = false;
        if (action.payload['Error Message'] || action.payload.message) {
          state.error = action.payload.message;
          state.companies = [];
        }
        state.companies = action.payload;
      })
      .addCase(getActiveCompanies.rejected, (state, action) => {
        state.loading = false;
        state.error = action.error.message;
      })
      .addCase(getCompanyDetail.pending, (state) => {
        state.loading = true;
      })
      .addCase(getCompanyDetail.fulfilled, (state, action) => {
        state.loading = false;
        [state.companyDetail] = action.payload;
      })
      .addCase(getCompanyDetail.rejected, (state, action) => {
        state.loading = false;
        state.error = action.error.message;
      })
      .addCase(getCompanyStatement.pending, (state) => {
        state.loading = false;
      })
      .addCase(getCompanyStatement.fulfilled, (state, action) => {
        state.loading = false;
        state.statement = action.payload;
      })
      .addCase(getCompanyStatement.rejected, (state, action) => {
        state.loading = false;
        state.error = action.error;
      });
  },
});

export const { filterCompanies, updateTyping } = companiesSlice.actions;
export default companiesSlice.reducer;
