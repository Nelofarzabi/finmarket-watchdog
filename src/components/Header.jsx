import React from 'react';
import { useDispatch } from 'react-redux';
import { filterCompanies, updateTyping } from '../redux/companies/companiesSlice';

const Header = () => {
  const dispatch = useDispatch();
  const handleChange = (e) => {
    dispatch(filterCompanies(e.target.value));
    dispatch(updateTyping(e.target.value));
  };
  return (
    <header className="flex justify-between items-center p-4 bg-[#2C3333] shadow-lg">
      <h2 className=" text-2xl text-[#F8F0E5]">Active Companies</h2>
      <form action="" className="w-[50%]">
        <input
          type="text"
          name=""
          id=""
          placeholder="search company by name..."
          className="border w-full px-2 py-1 rounded-md outline-none"
          onChange={handleChange}
        />
      </form>
    </header>
  );
};

export default Header;
