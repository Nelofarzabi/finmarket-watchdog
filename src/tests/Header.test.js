import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event'; // Import userEvent
import { useDispatch } from 'react-redux';
import { filterCompanies, updateTyping } from '../redux/companies/companiesSlice';
import Header from '../components/Header';

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
}));

describe('Header', () => {
  const mockDispatch = jest.fn();

  beforeEach(() => {
    useDispatch.mockReturnValue(mockDispatch);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('dispatches filterCompanies and updateTyping when input value changes', () => {
    const { getByPlaceholderText } = render(<Header />);
    const input = getByPlaceholderText('search company by name...');

    // Use userEvent to change the input value
    userEvent.type(input, 'Company');

    // Ensure the dispatch functions are called with the correct arguments
    expect(mockDispatch).toHaveBeenCalledWith(filterCompanies('Company'));
    expect(mockDispatch).toHaveBeenCalledWith(updateTyping('Company'));
  });

  test('matches snapshot', () => {
    const { container } = render(<Header />);

    // Check if the rendered component matches the snapshot
    expect(container).toMatchSnapshot();
  });
});
