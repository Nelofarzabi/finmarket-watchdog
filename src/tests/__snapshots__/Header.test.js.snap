// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header matches snapshot 1`] = `
<div>
  <header
    class="flex justify-between items-center p-4 bg-[#2C3333] shadow-lg"
  >
    <h2
      class=" md:text-2xl w-[50%] whitespace-nowrap text-xl text-[#F8F0E5]"
    >
      Active Companies
    </h2>
    <form
      action=""
      class="w-[50%]"
    >
      <input
        class="border w-full px-2 py-1 rounded-md outline-none"
        id=""
        name=""
        placeholder="search company by name..."
        type="text"
      />
    </form>
  </header>
</div>
`;
